### PR TITLE
FEAT: UIImage attachment tint

### LINKED ISSUE #34: [TextEffect] Support tint color for image attachments

### PR DESCRIPTION
**Pull request by jakubpetrik**
resolves #34

### CODE REVIEW

File: Styles/Classes/Extensions/UIImage+Attachment.swift [hunk 1]
```diff
@@ -5,7 +5,12 @@ import UIKit
 extension UIImage {
     func attachmentString(with style: TextStyle, baseAttributes: [NSAttributedStringKey: Any]) -> NSAttributedString {
         let attachment = NSTextAttachment()
-        attachment.image = self
+        let attributes = style.attributes
+        if let tintColor = attributes[.foregroundColor] as? UIColor, renderingMode != .alwaysOriginal {
```
Comment (radimhalfar) [inline Styles/Classes/Extensions/UIImage+Attachment.swift]: shouldn't it be invalid parameter if the rendering mode is set to always original ? Probably we should change it in this scope

File: Styles/Classes/Extensions/UIImage+Colorize.swift [hunk 2]
```diff
@@ -0,0 +1,71 @@
+// Source: https://gist.github.com/lynfogeek/4b6ce0117fb0acdabe229f6d8759a139
+
+import UIKit
+
+extension UIImage {
+
```
Comment (radimhalfar) [inline Styles/Classes/Extensions/UIImage+Colorize.swift]: line

File: Styles/Classes/Extensions/UIImage+Colorize.swift [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    // colorize image with given tint color
+    // this is similar to Photoshop's "Color" layer blend mode
+    // this is perfect for non-greyscale source images, and images that have both highlights and shadows that should be preserved
+    // white will stay white and black will stay black as the lightness of the image is preserved
+    func tint(tintColor: UIColor) -> UIImage {
+
```
Comment (radimhalfar) [inline Styles/Classes/Extensions/UIImage+Colorize.swift]: line

File: Styles/Classes/Extensions/UIImage+Colorize.swift [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        return modifiedImage { context, rect in
+            // draw black background - workaround to preserve color of partially transparent pixels
+            context.setBlendMode(.normal)
+            UIColor.black.setFill()
+            context.fill(rect)
+
+            // draw original image
+            context.setBlendMode(.normal)
+            context.draw(self.cgImage!, in: rect)
```
Comment (radimhalfar) [inline Styles/Classes/Extensions/UIImage+Colorize.swift]: can produce crash here

File: Styles/Classes/Extensions/UIImage+Colorize.swift [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+            // tint image (loosing alpha) - the luminosity of the original image is preserved
+            context.setBlendMode(.color)
+            tintColor.setFill()
+            context.fill(rect)
+
+            // mask by alpha values of original image
+            context.setBlendMode(.destinationIn)
+            context.draw(self.cgImage!, in: rect)
```
Comment (radimhalfar) [inline Styles/Classes/Extensions/UIImage+Colorize.swift]: also here, and the operation is for large images significantly time consuming

File: Styles/Classes/Extensions/UIImage+Colorize.swift [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    // fills the alpha channel of the source image with the given color
+    // any color information except to the alpha channel will be ignored
+    func fillAlpha(fillColor: UIColor) -> UIImage {
+
+        return modifiedImage { context, rect in
+            // draw tint color
+            context.setBlendMode(.normal)
+            fillColor.setFill()
+            context.fill(rect)
+            //            context.fillCGContextFillRect(context, rect)
```
Comment (radimhalfar) [inline Styles/Classes/Extensions/UIImage+Colorize.swift]: forgotten comment

File: Styles/Classes/Extensions/UIImage+Colorize.swift [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+            // mask by alpha values of original image
+            context.setBlendMode(.destinationIn)
+            context.draw(self.cgImage!, in: rect)
+        }
+    }
+
+    // fills the alpha channel of the source image with the given color
+    // any color information except to the alpha channel will be ignored
+    func fillAlpha(fillColor: UIColor) -> UIImage {
+
```
Comment (radimhalfar) [inline Styles/Classes/Extensions/UIImage+Colorize.swift]: line

File: Styles/Classes/Extensions/UIImage+Colorize.swift [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+            //            context.fillCGContextFillRect(context, rect)
+
+            // mask by alpha values of original image
+            context.setBlendMode(.destinationIn)
+            context.draw(self.cgImage!, in: rect)
+        }
+    }
+
+    private func modifiedImage( draw: (CGContext, CGRect) -> ()) -> UIImage {
+
```
Comment (radimhalfar) [inline Styles/Classes/Extensions/UIImage+Colorize.swift]: line

File: Styles/Classes/Extensions/UIImage+Colorize.swift [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+            context.fill(rect)
+            //            context.fillCGContextFillRect(context, rect)
+
+            // mask by alpha values of original image
+            context.setBlendMode(.destinationIn)
+            context.draw(self.cgImage!, in: rect)
+        }
+    }
+
+    private func modifiedImage( draw: (CGContext, CGRect) -> ()) -> UIImage {
```
Comment (radimhalfar) [inline Styles/Classes/Extensions/UIImage+Colorize.swift]: preffer Void over ()

File: Styles/Classes/Extensions/UIImage+Colorize.swift [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+        return modifiedImage { context, rect in
+            // draw tint color
+            context.setBlendMode(.normal)
+            fillColor.setFill()
+            context.fill(rect)
+            //            context.fillCGContextFillRect(context, rect)
+
+            // mask by alpha values of original image
+            context.setBlendMode(.destinationIn)
+            context.draw(self.cgImage!, in: rect)
```
Comment (radimhalfar) [inline Styles/Classes/Extensions/UIImage+Colorize.swift]: can produce crash

File: Styles/Classes/Extensions/UIImage+Colorize.swift [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+        let rect = CGRect(x: 0.0, y: 0.0, width: size.width, height: size.height)
+
+        draw(context, rect)
+
+        let image = UIGraphicsGetImageFromCurrentImageContext()
+        UIGraphicsEndImageContext()
+        return image!
+    }
+
```
Comment (radimhalfar) [inline Styles/Classes/Extensions/UIImage+Colorize.swift]: line

File: Styles/Classes/Extensions/UIImage+Colorize.swift [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+        context.translateBy(x: 0, y: size.height)
+        context.scaleBy(x: 1.0, y: -1.0)
+
+        let rect = CGRect(x: 0.0, y: 0.0, width: size.width, height: size.height)
+
+        draw(context, rect)
+
+        let image = UIGraphicsGetImageFromCurrentImageContext()
+        UIGraphicsEndImageContext()
+        return image!
```
Comment (radimhalfar) [inline Styles/Classes/Extensions/UIImage+Colorize.swift]: can produce crash

File: Styles/Classes/Extensions/UIImage+Attachment.swift (hunk 1 above)
Comment (jakubpetrik) [reply Styles/Classes/Extensions/UIImage+Attachment.swift]: The idea was to tint image only if that is intented to do so. You can use some text style that has .foregroundColor which you don't want and bunch of other attributes that you are interested in but you don't want image to be tinted. In this case we choose to tint based on image rendering mode. 
So a default rendering mode for image attachment is to tint which means that we tint any image that has rendering mode `default` or `alwaysTemplate`.

File: Styles/Classes/Extensions/UIImage+Colorize.swift (hunk 2 above)
Comment (jakubpetrik) [reply Styles/Classes/Extensions/UIImage+Colorize.swift]: This is not our source.

File: Styles/Classes/Extensions/UIImage+Colorize.swift (hunk 3 above)
Comment (jakubpetrik) [reply Styles/Classes/Extensions/UIImage+Colorize.swift]: This is not our source.

**Pull request merged**